File: Python/Algorithm/DFS.py
# Complexity: O(V + E)
# eg = Number of edges, vr = Number of vertices

def dfs(par):
    vis[par] = True
    for cld in gp[par]:
        if not vis[cld]: dfs(cld)

# Operation
global gp, vis
vr, eg = map(int, input().split())
gp = {i: [] for i in range(vr)}
vis = [False] * vr

for _ in range(eg):
    v1, v2 = map(int, input().split())
    gp[v1].append(v2); gp[v2].append(v1)

dfs(1)

File: CPP/Algorithm/DFS.cpp
// Complexity: O(V + E)
https://cses.fi/problemset/task/1666

vector<int> gp[N]; bool vis[N];
void dfs(int par = 1) {
	vis[par] = true;
	for (auto &cld : gp[par]) {
		if (!vis[cld]) dfs(cld);
	}
}

// Operation:
	int vr, eg; cin >> vr >> eg;
	for (i = 0; i < eg; ++i) {
		int v1, v2; cin >> v1 >> v2;
		gp[v1].push_back(v2); gp[v2].push_back(v1);
	}
	vector<int> com;
	for (i = 1; i <= vr; ++i) {
		if (!vis[i]) {
			com.push_back(i); dfs(i);
		}
	}
    n = com.size(); cout << n - 1 << '\n';
	for (i = 0; i < n - 1; ++i) cout << com[i] << ' ' << com[i + 1] << '\n';